File: backend/app/schemas/personal_files.py
"""数据格式层：定义个人文件分页、原件详情和行程版本详情。"""

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field

from app.schemas.attachment import AttachmentView
from app.schemas.itinerary import PlanSnapshot
from app.schemas.requirement.base import TravelRequestExtraction

FileKind = Literal["attachment", "itinerary"]


class PersonalFileItem(BaseModel):
    """文件条目类：记录原有文件身份和来源，行程大小在生成下载内容前保持未知。"""

    id: UUID
    kind: FileKind
    file_name: str
    mime_type: str
    size_bytes: int | None
    session_id: UUID
    session_title: str
    created_at: datetime
    version: int | None = None
    version_count: int | None = None


class PersonalFilePage(BaseModel):
    """文件分页类：总数由数据库按相同过滤条件计算，不受当前页大小影响。"""

    items: list[PersonalFileItem]
    total: int
    offset: int
    limit: int


class PersonalFileDetail(BaseModel):
    """文件详情类：原件复用附件信息，行程复用已保存的版本和完整需求快照。"""

    item: PersonalFileItem
    attachment: AttachmentView | None = None
    itinerary: PlanSnapshot | None = None
    extraction: TravelRequestExtraction | None = None
    versions: list[PersonalFileItem] = Field(default_factory=list)

File: frontend/src/api/personalFiles.ts
/** 个人文件接口层：跨会话查询当前账号文件，详情和下载仍由服务端逐次检查归属。 */
import type { AttachmentView } from './attachments'
import { attachmentContentUrl } from './attachments'
import type { PlanSnapshot, TravelRequirement } from './requirements'
import { apiFetch, readResponse } from './http'

export type PersonalFileKind = 'attachment' | 'itinerary'
export interface PersonalFileItem {
  id: string; kind: PersonalFileKind; file_name: string; mime_type: string
  size_bytes: number | null; session_id: string; session_title: string; created_at: string
  version: number | null; version_count: number | null
}
export interface PersonalFilePage { items: PersonalFileItem[]; total: number; offset: number; limit: number }
export interface PersonalFileDetail {
  item: PersonalFileItem; attachment: AttachmentView | null
  itinerary: PlanSnapshot | null; extraction: TravelRequirement | null; versions: PersonalFileItem[]
}

/** 分页查询函数：不把全部文件先下载到浏览器，筛选与总数都来自数据库。 */
export async function listPersonalFiles(options: {
  sessionId?: string | null; kind?: PersonalFileKind | ''; q?: string; offset?: number; limit?: number
}): Promise<PersonalFilePage> {
  const query = new URLSearchParams({ offset: String(options.offset ?? 0), limit: String(options.limit ?? 30) })
  if (options.sessionId) query.set('session_id', options.sessionId)
  if (options.kind) query.set('kind', options.kind)
  if (options.q?.trim()) query.set('q', options.q.trim())
  return readResponse(await apiFetch(`/api/v1/personal-files?${query}`, { cache: 'no-store' }))
}

/** 文件详情函数：版本编号始终使用行程UUID，不用标题或文件名匹配。 */
export async function getPersonalFile(item: Pick<PersonalFileItem, 'id' | 'kind'>): Promise<PersonalFileDetail> {
  return readResponse(await apiFetch(`/api/v1/personal-files/${item.kind}/${encodeURIComponent(item.id)}`, { cache: 'no-store' }))
}

/** 下载地址函数：原件沿用原有受保护地址，行程由服务器按所选版本即时生成。 */
export function personalFileDownloadUrl(item: PersonalFileItem): string {
  return item.kind === 'attachment' ? attachmentContentUrl(item.session_id, item.id)
    : `/api/v1/personal-files/itinerary/${encodeURIComponent(item.id)}/download`
}
